### DISCUSSION TITLE
Running a custom theia image(coffee-editor) in theia-cloud

### DISCUSSION THREAD

**Discussion by Deel96** [eclipse-theia/theia-cloud#74] [Q&A] [answered]
Hi,

I setup theia-cloud  according to the getting started guide for minikube (which was really straightforward btw, kudos for that!).
As a next step I want to deploy a custom theia version (e.g. the coffee editor example)
After my research I have to edit this line:
https://github.com/eclipsesource/theia-cloud/blob/main/helm/theia.cloud/values.yaml#L21

But I still get the blueprint theia.
After my research the published port should be the same like in the example (3000)

Is it somewhere documented what has to be done to switch the blueprint theia for another ide like the coffee-editor example?

**Comment by jfaltermeier** [comment, marked as answer]
Hi, 

thank you. 

The minikube docs use this values file as part of the helm install: valuesMinikube.yaml. 
So adjusting the image here should work: https://github.com/eclipsesource/theia-cloud/blob/0f8d86487f84b0d7d54b9d9ab82fc3f600512d46/helm/theia.cloud/valuesMinikube.yaml#L10

You might also have to adjust the user id as well: https://github.com/eclipsesource/theia-cloud/blob/0f8d86487f84b0d7d54b9d9ab82fc3f600512d46/helm/theia.cloud/templates/theia-appdefinition-spec.yaml#L9
Maybe there is no user with uid 101 in your custom image which might lead to errors, where files can't be read. 

After installing Theia.cloud running `kubectl get appdefinition theia-cloud-demo -n theiacloud -o yaml` should show you the app definition with all the used values including image.

**Comment by Deel96** [comment]
Hi @jfaltermeier,

Changing the image at `valuesMinikube.yaml` shows me that the new image is being used.

`kubectl get appdefinition theia-cloud-demo -n theiacloud -o yaml`
<details>

```yaml

apiVersion: theia.cloud/v1beta
kind: AppDefinition
metadata:
  annotations:
    meta.helm.sh/release-name: theia-cloud
    meta.helm.sh/release-namespace: theiacloud
  creationTimestamp: "2022-08-22T19:23:12Z"
  generation: 1
  labels:
    app.kubernetes.io/managed-by: Helm
  name: theia-cloud-demo
  namespace: theiacloud
  resourceVersion: "5416"
  uid: 968a38f3-a37e-4062-8c89-54bcf8305774
spec:
  downlinkLimit: 30000
  host: ws.192.168.49.2.nip.io
  image: browork/ide:latest
  ingressname: theia-cloud-demo-ws-ingress
  limitsCpu: "2"
  limitsMemory: 1200M
  maxInstances: 10
  minInstances: 0
  name: theia-cloud-demo
  port: 3000
  requestsCpu: 100m
  requestsMemory: 1000M
  uid: 1000
  uplinkLimit: 30000
```
</details>

If I run my image locally the user has the uid of 1000, so I changed it also in `theia-appdefinition-spec.yaml`

If someone in the future comes across this post:
Since my image is pretty big (3GB) I recommend prefetching it in minikube:
```
eval $(minikube -p minikube docker-env)
docker pull browork/ide:latest
```

With the changes above a new workspace starts almost immediately. Thank you very much!

Maybe a bit off-topic:
I tried to deploy the same image (relative old vanilla version of the coffee-editor) on eclipse che. On eclipse-che I get the frontend  "loaded" but everything is blank.
Custom editors like tree-editor or workflow editor are blank because of a (presumably) missing backend connection. 
(errors about connection to localhost:8081/api/v1/server/configure were refused. I saw that the sourcecode has localhost hard coded in many places).

But when deploying via theia-cloud with the same code/image it works out of the box (Mostly, sometimes have to refresh multiple times to get everything working as expected.
Both running on minikube.

Do you have a clue why it works with theia-cloud out of the box?
Strangely enough the icons for .coffee etc files are present in che, but not in the theia-cloud deployment.

**Reply by jfaltermeier** [reply]
Hard to tell why it is not working in Che. Maybe the server running on port 8081 is taking too long to launch or crashes. If the terminal is working in Theia, you might run `ps -aux` to check if there are java processes for the server

**Reply by Deel96** [reply]
ps -aux showed that the model and glsp server were not running. The modelserver gave an error with little information. The reason at the end was, that the ide container ran out of resourced. Rising cpu and memory limits solved the issue.
